### PR TITLE
WIP: Moodle on NGINX not Apache

### PR DESCRIPTION
**Pull request by holta**
A possible tightening up, building upon @jvonau's PR #2839, after much testing this morning.

Refs:

- #2762
- PR #2784
- #2785
- #2829

### CODE REVIEW

**Comment by jvonau** [comment]
If you are the nginx expert why was the PR not started by yourself? Do you know what the differences are between fastcgi_params fastcgi.conf? PS the apt list was taken straight from https://github.com/moodlebox/moodlebox/blob/master/roles/packages/vars/main.yml guess you know better than the official release manager of moodlebox.

**Comment by jvonau** [comment]
clean box testing required, I do not fully understand why the moodle-nginx.conf was [not working](https://github.com/iiab/iiab/pull/2839#issuecomment-872527111) until [maps](https://github.com/iiab/iiab/pull/2839#issuecomment-872538296) was installed.

**Comment by holta** [comment]
> Do you know what the differences are between fastcgi_params fastcgi.conf?

Just FYI both work, but fastcgi.conf is a gratuitous shortcut (compared to fastcgi_params) that adds 1 unneeded line:

```
root@u21-srv:/etc/nginx# diff fastcgi_params fastcgi.conf
1a2
> fastcgi_param  SCRIPT_FILENAME    $document_root$fastcgi_script_name;
```

i.e. this exact same line is already included in `/etc/ninx/conf.d/moodle-nginx.conf` -- and perhaps more important it's part of the [official Moodle-with-NGINX instructions](https://docs.moodle.org/311/en/Nginx) -- so probably better to stick with the most recognizable NGINX/Moodle norms, that more people are likely to recognize and support long-term.

**Comment by holta** [comment]
> the apt list was taken straight from https://github.com/moodlebox/moodlebox/blob/master/roles/packages/vars/main.yml

Really great find.

I've done a careful reconciliation between the 2, and many clean box installs.

And will also continue the experimentation.

**Comment by jvonau** [comment]
Figure out how 'root' works and if we are abusing or misusing the directive
ubuntu@box:/opt/iiab/iiab$ grep -r root roles/*/templates/*nginx*
```
roles/mediawiki/templates/mediawiki-nginx.conf.j2:# this works if (docroot)/{{ mediawiki_symlink }} links to install
roles/mediawiki/templates/mediawiki-nginx.conf.j2:		fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
roles/mediawiki/templates/mediawiki-nginx.conf.j2:#		fastcgi_param SCRIPT_FILENAME $document_root/{{ mediawiki_symlink }}/mw-config/$fastcgi_script_name;
roles/mediawiki/templates/mediawiki-nginx.conf.j2:# Explicit access to the root website, redirect to main page (adapt as needed)
roles/moodle/templates/moodle-nginx.conf.j2:location ~ ^/moodle/dataroot/ {
roles/moodle/templates/moodle-nginx.conf.j2:    fastcgi_param   SCRIPT_FILENAME   $document_root/$fastcgi_script_name;
roles/moodle/templates/moodle-nginx.conf.j2:    root {{ iiab_base }};
roles/nextcloud/templates/nextcloud-nginx.conf.j2:# https://docs.nextcloud.com/server/18/admin_manual/installation/nginx.html#nextcloud-in-the-webroot-of-nginx
roles/nextcloud/templates/nextcloud-nginx.conf.j2:    # Path to the root of your installation
roles/nextcloud/templates/nextcloud-nginx.conf.j2:    #root {{ nextcloud_root_dir }};    # /library/www/nextcloud IN FUTURE ??
roles/nextcloud/templates/nextcloud-nginx.conf.j2:    root {{ nextcloud_base_dir }};    # /library/www == {{ doc_base }} FOR NOW
roles/nextcloud/templates/nextcloud-nginx.conf.j2:        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
roles/nginx/templates/admin-console-nginx.conf.unused:    fastcgi_param   SCRIPT_FILENAME    $document_root$fastcgi_script_name;
roles/nginx/templates/kalite-nginx.conf.unused:    set $kalite_home {{ kalite_root }};
roles/nginx/templates/nginx.conf.j2:    log_format scripts '$request > $document_root$fastcgi_script_name $fastcgi_path_info';
roles/osm-vector-maps/templates/osm-vector-maps-nginx.conf.j2:    fastcgi_param   SCRIPT_FILENAME    $document_root$fastcgi_script_name;
roles/osm-vector-maps/templates/osm-vector-maps-nginx.conf.j2:   root /library/www;
roles/wordpress/templates/wordpress-nginx.conf.j2:    root {{ content_base }};
roles/wordpress/templates/wordpress-nginx.conf.j2:          fastcgi_param   SCRIPT_FILENAME    $document_root$fastcgi_script_name;          
```

**Comment by holta** [comment]
> Figure out how 'root' works and if we are abusing or misusing the directive

Your use of the `root` directive looks good to me.

I researched that earlier this morning.  It's basically a more concise version of the `alias` directive.

**Comment by holta** [comment]
> many clean box installs

i.e. this all looks great on PHP 7.4 (Ubuntu 21.04) and 8.0 (Ubuntu 21.10) &mdash; thanks to @jvonau and Moodlebox.net for finally making this critical advance happen.

I'll also do PHP 7.3 (32-bit RaspiOS Lite) testing shortly.

**Comment by holta** [comment]
> I'll also do PHP 7.3 (32-bit RaspiOS Lite) testing shortly.

Done: this PR should be ready to merge toward expanding community testing of http://box.lan/moodle hopefully starting today or very soon!

This PR and associated sanity checks over the past 48h are a giant advance in:

1. Tightening up and verifying Moodle functionality ~5 years after it was introduced to IIAB.
2. Bringing IIAB's [2-year transition from Apache to NGINX](https://github.com/iiab/iiab/tree/master/roles/nginx) to a close very shortly now.

Profound thanks to @jvonau & @martignoni for making this happen!  @martignoni would you have time to give this PR a quick review?  _Many thanks if possible!_

**Comment by jvonau** [comment]
> 
> Really great find.
> 

Not the first time I've posted the link. Here is their [nginx config](https://github.com/moodlebox/moodlebox/blob/master/roles/webserver/templates/etc/nginx/sites-available/default.j2), the other major source used was osm-vector-maps  

> I've done a careful reconciliation between the 2, and many clean box installs.
>

And you have not been able to reproduce the apparent need for maps to be installed also? https://github.com/iiab/iiab/pull/2839#issuecomment-872527111, and in hindsight looks like it was the trailing slash issue.
> And will also continue the experimentation.

> > Do you know what the differences are between fastcgi_params fastcgi.conf?
> 
> Just FYI both work, but fastcgi.conf is a gratuitous shortcut (compared to fastcgi_params) that adds 1 unneeded line:
> 
> ```
> root@u21-srv:/etc/nginx# diff fastcgi_params fastcgi.conf
> 1a2
> > fastcgi_param  fastcgi_param   SCRIPT_FILENAME  $document_root/$fastcgi_script_name;    $document_root$fastcgi_script_name;
> ```
> 
> i.e. this exact same line is already included in `/etc/ninx/conf.d/moodle-nginx.conf` -- and perhaps more important it's part of the [official Moodle-with-NGINX instructions](https://docs.moodle.org/311/en/Nginx) -- so probably better to stick with the most recognizable NGINX/Moodle norms, that more people are likely to recognize and support long-term.

Yea, I tried mediawiki's approach of using a symlink in the webroot path, the path to the script would of needed to be altered. When I got the site to come up by hand was when I went to template the config, as a starting point. Wasn't that hard, just forgot to undo that change. The relationship between '$document_root' and 'root' was what I was exploring, if 'root' was 'alias' then the   'SCRIPT_FILENAME might be {{ moodle_base }}/$fastcgi_script_name;' but I didn't try that yet. More I want to know why it works as is and wanted to get more eyes on the subject of 'root' vs 'alias' and their correct use because if appears we use 'root' in a few places. I'm thinking when the request comes in for http://box.lan/moodle, nginx puts the path together as $document_root/moodle with 'root /opt/iiab' being used directly in $document_root which is the path on the filesystem. Should that be correct porting the rest over to nginx should be pretty straight forward. Just want to fully understand nginx's handling so we are not going down the wrong path before looking at Elgg.

**Comment by holta** [comment]
> Just want to fully understand nginx's handling so we are not going down the wrong path before looking at Elgg.

ASIDE: I get the feeling IIAB's Elgg LTS is broken (right after login) even with Apache.

This is the LTS version (2.3.17) in a holding pattern from years ago FWIW, with 2.3.x having been released in 2016.

Even the non-LTS version (3.3.18) — which was never integrated with IIAB — is likely to be replaced by [Elgg 4.0.0](https://elgg.org/about/download) (active Beta) shortly.

At a higher level, I also have a concern that Elgg is veritably no longer in broadly popular use.  Now that these chat-like features have been wrapped into larger LMS-like products.

File: roles/moodle/templates/moodle-nginx.conf.j2 [hunk 1]
```diff
@@ -1,13 +1,32 @@
-location /moodle {
-    proxy_set_header X-Real-IP  $remote_addr;
-    proxy_set_header X-Forwarded-For $remote_addr;
-    proxy_set_header Host $host;
-    proxy_pass http://127.0.0.1:{{ apache_port }};
+# 2021-07-02: Let's monitor & learn from these 2 pages year-by-year:
+# https://docs.moodle.org/311/en/Nginx
+# https://github.com/moodlebox/moodlebox/blob/master/roles/webserver/templates/etc/nginx/sites-available/default.j2
+
+# Odd that MoodleBox.net is using only the 2nd half of https://docs.moodle.org/311/en/Nginx#XSendfile_aka_X-Accel-Redirect ?
```
Comment (martignoni) [inline roles/moodle/templates/moodle-nginx.conf.j2]: If I understand correctly the comment: First half is done here https://github.com/moodlebox/moodlebox/blob/851d1eb2fd3672e275ca2512eabf7eeed79efe2b/roles/moodleinstall/tasks/coreinstall.yml#L102-L103

File: roles/moodle/templates/moodle-nginx.conf.j2 [hunk 2]
```diff
@@ -1,13 +1,32 @@
...
+#location ~ ^/moodle/dataroot/ {
+#    internal;
+#    alias {{ moodle_data }};
```
Comment (martignoni) [inline roles/moodle/templates/moodle-nginx.conf.j2]: As stated in Moodle documentation, ensure this path (ie. `{{ moodle_data }}`) ends with a slash

**Comment by martignoni** [comment]
> @martignoni would you have time to give this PR a quick review? _Many thanks if possible!_

Done. Seems all good to me but have a look at my comments above.

File: roles/moodle/templates/moodle-nginx.conf.j2 (hunk 1 above)
Comment (jvonau) [reply roles/moodle/templates/moodle-nginx.conf.j2]: Could use this link for our install.yml?

File: roles/moodle/templates/moodle-nginx.conf.j2 (hunk 2 above)
Comment (jvonau) [reply roles/moodle/templates/moodle-nginx.conf.j2]: yup noted https://docs.moodle.org/311/en/Nginx

File: roles/moodle/templates/moodle-nginx.conf.j2 (hunk 1 above)
Comment (martignoni) [reply roles/moodle/templates/moodle-nginx.conf.j2]: Yes (as far as i'm concerned)

**Comment by holta** [comment]
Many thanks to all who've contributed over the years (@georgejhunt and others worked on this in [2015 and 2016](https://github.com/XSCE/xsce/commits/master/roles/moodle), but also many others prior!)

I'd like to merge this very shortly now &mdash; it's been tested on RaspiOS Lite (PHP 7.3, RPi 4) & Ubuntu Server 21.10 pre-release (PHP 8.0, VM) and ready for much broader community testing.

Finally, @tim-moody and others may have suggestions regarding the security of `/library/moodle` discussed here: 
https://github.com/holta/iiab/blob/6ec9fe8a5e220bb417d7e0195be2e32a36a71777/roles/moodle/tasks/install.yml#L117-L125

**Comment by holta** [comment]
@deldesir has agreed to review this Moodle/NGINX integration, even if it's later this week after the PR is merged.

**Comment by holta** [comment]
Also smoke-tested on 64-bit Ubuntu Server 21.04 (PHP 7.4, RPi 4).  So this PR will now be merged.

Additional testing on Debian 11[*] would be great in coming days if anybody can!

_[*] Or on a pre-release of Raspberry Pi OS 11, if you can find one!_

**Pull request merged**